### PR TITLE
Integrated SSMA aggregation

### PR DESCRIPTION
**Pull request by AlmogDavid**
Integrated the "Sequential Signal Mixing Aggregation" (SSMA) aggregation method from: https://arxiv.org/abs/2409.19414

### CODE REVIEW

**Comment by Kh4L** [comment]
Looks great! Did you try it out with some OGB benchmarks?

I think it'd be nice to add an option to try it out in [ogbn_train.py ](https://github.com/pyg-team/pytorch_geometric/blob/f5c829344517c823c24abb08ce2fc7cf00ff29f7/examples/ogbn_train.py#L22) - in a followup PR, and see if we get the  accuracy boost claimed in Table 2 ( like the 15-20% boost in ogbn-products :rocket: )!

**Comment by AlmogDavid** [comment]
> Looks great! Did you try it out with some OGB benchmarks from PyG examples?
> 
> I think it'd be nice to add an option to try it out in [ogbn_train.py ](https://github.com/pyg-team/pytorch_geometric/blob/f5c829344517c823c24abb08ce2fc7cf00ff29f7/examples/ogbn_train.py#L22) - in a followup PR, and see if we get the accuracy boost claimed in Table 2 ( like the 15-20% boost in ogbn-products 🚀 )!

Thanks!
You can find the experimental setting and to reproduce the results using our repository: https://almogdavid.github.io/SSMA/
There you can find full set of datasets and configurations to try out yourself, not sure if its appropriate to link to the repository from the documentation itself, but if you think it will be helpful I will do it.

**Comment by AlmogDavid** [comment]
@wsad1 , @EdisonLeeeee . Is there some ETA for a review?

Thanks.

**Comment by AlmogDavid** [comment]
@wsad1 , @EdisonLeeeee - Any updates?

**Comment by AlmogDavid** [comment]
@rusty1s  - How can we progress this MR?

File: torch_geometric/nn/aggr/ssma.py [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+                          self._max_neighbors)  # [N, #groups, #neighbors]
+
+        # Compute softmax over the neighbors based on the attention weights
+        # and the graph topology
+        edge_attention_ste = softmax(
+            F.leaky_relu(x_l[edge_index[0]] + x_r[edge_index[1]]) /
+            self._neighbor_att_temp, edge_index[1])  # [E, #groups, #neighbors]
+        return edge_attention_ste
+
+    def pre_aggregation_hook(self, module, inputs):
```
Comment (wsad1) [inline torch_geometric/nn/aggr/ssma.py:114]: Could you explain why the prehook is needed?
Maybe add to the documentation why this pre hook is needed, and why these opreations can't be implemented as part of the forward pass.

File: torch_geometric/nn/aggr/ssma.py [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+from torch import Tensor
+
+from torch_geometric.nn.aggr import Aggregation
+from torch_geometric.utils import scatter, softmax
+
+
+class SSMAAggregation(Aggregation):
+    r"""Performs the Sequential Signal Mixing Aggregation (SSMA) method from
+    `"Sequential Signal Mixing Aggregation for Message Passing Graph
+    Neural Networks" <https://arxiv.org/abs/2409.19414>`_ paper.
```
Comment (wsad1) [inline torch_geometric/nn/aggr/ssma.py:16]: Add the forumulae for the aggregation , like we have for other aggregators https://pytorch-geometric.readthedocs.io/en/latest/generated/torch_geometric.nn.aggr.PowerMeanAggregation.html.

Comment (AlmogDavid) [reply torch_geometric/nn/aggr/ssma.py:16]: Done

File: torch_geometric/nn/aggr/ssma.py (hunk 1 above)
Comment (AlmogDavid) [reply torch_geometric/nn/aggr/ssma.py:120]: Thanks for your comment.

The pre-hook is needed only when using the attention mechanism in SSMA. This function creates the edge weights for attention, which requires edge information (source + target). Since this information is unavailable in the forward pass, the pre-hook is necessary.

Based on your feedback:
	•	Added documentation in the code.
	•	Ensured the pre-hook is called **only** when use_attention == True.
	•	Updated tests to verify this case.

Let me know if further clarification is needed.

**Comment by AlmogDavid** [comment]
@wsad1 thanks for the review, I addressed all of you comments.
Please let me know if more things are needed. Thanks.

File: torch_geometric/nn/aggr/ssma.py [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+                          self._max_neighbors)  # [N, #groups, #neighbors]
+
+        # Compute softmax over the neighbors based on the attention weights
+        # and the graph topology
+        edge_attention_ste = softmax(
+            F.leaky_relu(x_l[edge_index[0]] + x_r[edge_index[1]]) /
+            self._neighbor_att_temp, edge_index[1])  # [E, #groups, #neighbors]
+        return edge_attention_ste
+
+    def pre_aggregation_hook(self, module, inputs):
```
Comment (wsad1) [inline torch_geometric/nn/aggr/ssma.py:120]: The hook will only get the arguments `forward` (gets)[https://pytorch.org/docs/stable/generated/torch.nn.modules.module.register_module_forward_hook.html]. So the hook won't have access to `edge_index`.

Comment (wsad1) [reply torch_geometric/nn/aggr/ssma.py:120]: So Instead of making this a prehook, make this a static function that the user could call before calling forward.

**Review by wsad1 — COMMENTED** [review]
Thanks for the contribution. 
I think its best to move this to [torch_geometric/contrib](https://github.com/pyg-team/pytorch_geometric/tree/master/torch_geometric/contrib) folder.

File: torch_geometric/nn/aggr/ssma.py (hunk 3 above)
Comment (AlmogDavid) [reply torch_geometric/nn/aggr/ssma.py]: Thanks for the review. This is not a forward hook. it's a pre aggregation hook (propagate_forward_pre_hook: https://pytorch-geometric.readthedocs.io/en/latest/generated/torch_geometric.nn.conv.MessagePassing.html?highlight=propagate_forward_pre_hook#torch_geometric.nn.conv.MessagePassing.register_propagate_forward_pre_hook) It does have access to edge index.

It also modifies private fields in the aggregation so I don't see how it can become a static method.

**Comment by AlmogDavid** [comment]
Moved the aggregation to the contrib folder as suggested.

**Comment by AlmogDavid** [comment]
@rusty1s What can we do to move this MR forward? I’ve addressed all @wsad1’s comments and implemented the requested changes. I also reached out to @EdisonLeeeee, but he doesn’t have time for a code review right now.

**Comment by AlmogDavid** [comment]
Someone, any updates? @wsad1 @rusty1s

File: test/contrib/nn/aggr/test_ssma.py [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+
+def test_representation():
+    ssma_agg = SSMAAggregation(in_dim=10, num_neighbors=3, mlp_compression=0.1,
+                               use_attention=False, n_heads=59, temp=0.2,
+                               att_feature="dummy", learn_affine=True)
+
+    rep = str(ssma_agg)
+    gen_rep = eval(rep)
+
+    assert str(gen_rep) == rep
```
Comment (wsad1) [inline test/contrib/nn/aggr/test_ssma.py:97]:
```suggestion
```

**Review by wsad1 — APPROVED** [review]

**Comment by AlmogDavid** [comment]
@rusty1s - Can you please review?

**Comment by AlmogDavid** [comment]
@wsad1 @rusty1s  ?

**Comment by AlmogDavid** [comment]
@akihironitta Can we merge?

**Comment by AlmogDavid** [comment]
@akihironitta  @wsad1 - The MR is open for 9 months...
What can we do in order to push it?

**Comment by AlmogDavid** [comment]
@akihironitta ....
